sale_coupon_advanced/models/sale_coupon.py

```diff
@@ -1,9 +1,88 @@
 # Copyright 2020 Camptocamp SA
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
-from odoo import fields, models
+from odoo import _, api, exceptions, fields, models
 
 
 class SaleCouponProgram(models.Model):
     _inherit = "sale.coupon.program"
 
     is_cumulative = fields.Boolean(string="None-cumulative Promotion")
+
+    # Add possibility to use discount only on first order of a customer
+    first_order_only = fields.Boolean(
+        string="Apply only first",
+        help="Apply only on the first order of each client matching the conditions",
+    )
+
+    first_n_orders_only = fields.Integer(
+        help="Maximum number of sales orders of the customer in which reward \
+         can be provided",
+        string="Apply only on the next ",
+        default=0,
+    )
+
+    @api.constrains("first_n_orders_only")
+    def _constrains_first_n_orders_positive(self):
+        for record in self:
+            if record.first_order_only < 0:
+                raise exceptions.ValidationError(
+                    _("`Apply only on the next` should not be a negative value.")
+                )
+
+    def _get_order_count(self, order):
+        return self.env["sale.order"].search_count(
+            [
+                ("partner_id", "=", order.partner_id.id),
+                ("state", "!=", "cancel"),
+                ("id", "!=", order.id),
+            ]
+        )
+
+    def _check_promo_code(self, order, coupon_code):
+        order_count = self._get_order_count(order)
+        if self.first_order_only and order_count:
+            return {"error": _("Coupon can be used only for the first sale order!")}
+        if self.first_n_orders_only and order_count >= self.first_n_orders_only:
+            return {
+                "error": _(
+                    "Coupon can be used only for the first %s sale order!"
+                    % (str(self.first_n_orders_only))
+                )
```

Comment:
Be carefull to not include context values in a translatable field and use a different string terminal formatter instead of a string convert:
```suggestion
                "error": _("Coupon can be used only for the first %d sale order!")
                            % self.first_n_orders_only
```